### PR TITLE
ARROW-11049: [Python] Expose alternate memory pools

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

File: python/pyarrow/memory.pxi [hunk 1]
```diff
@@ -127,6 +127,44 @@ def logging_memory_pool(MemoryPool parent):
...
+        MemoryPool pool = MemoryPool.__new__(MemoryPool)
+    pool.init(c_system_memory_pool())
+    return pool
+
+
+def jemalloc_memory_pool():
+    """
+    Return a memory pool based on the jemalloc heap.
+
+    NotImplemented is raised if this Arrow build does not enable jemalloc.
```
Comment (kszucs) [inline python/pyarrow/memory.pxi:144]: Perhaps you should add a bit more context about where we use the allocators. Like jemalloc is used and enabled in our packages.

Comment (pitrou) [reply python/pyarrow/memory.pxi:144]: I don't think jemalloc is enabled on all platforms, is it?

Comment (kszucs) [reply python/pyarrow/memory.pxi:144]: *unix packages

Comment (pitrou) [reply python/pyarrow/memory.pxi:144]: Well, we'll certain forgot to update this docstring if we change the build settings, so I'd rather not be more precise.

Comment (kszucs) [reply python/pyarrow/memory.pxi:144]: Right

File: python/pyarrow/tests/test_memory.py [hunk 2]
```diff
@@ -133,3 +133,24 @@ def test_env_var():
     if should_have_mimalloc:
         check_env_var("mimalloc", ["mimalloc"])
     check_env_var("nonexistent", possible_backends, expect_warning=True)
+
+
+def test_specific_memory_pools():
```
Comment (kszucs) [inline python/pyarrow/tests/test_memory.py:138]: Will this case fail if we build a wheel without the specific optional allocator enabled (based on the should have flag [definitions](https://github.com/apache/arrow/blob/053f4742c7fde9c1e3b3a0858c4de6dc05efec98/python/pyarrow/tests/test_memory.py#L29-L30))?

If yes we may want to find a better way to retrieve the built allocator in a follow-up.

Comment (pitrou) [reply python/pyarrow/tests/test_memory.py:138]: Unless the case happens, I don't think we should worry.

**Comment by jorisvandenbossche** [comment]
Looks good, no comments on the Python code. Maybe not critical for *this* PR, but I have some suggestions (or questions) for expanded docs:

- I would add some more text to the `python/memory.rst`. There is now "PyArrow uses a default built-in memory pool, but depending on how Arrow was built, there may be additional memory pools to choose from."  
  -> Could mention which different options are possible (system, jemalloc, mimalloc), referencing the API page with the overview of the related functions  
  -> Is there a way to see interactively which one is being used?
  -> Can you switch the pool being used? (eg do `pa.set_memory_pool(pa.mimalloc_memory_pool())` ?

**Comment by pitrou** [comment]
Ok, I've improved the docs a bit. Will merge if green.

**Comment by jorisvandenbossche** [comment]
Thanks for the doc update!

**Pull request closed without merging**